hack/generator/pkg/codegen/pipeline_improve_resource_pluralization.go

```diff
@@ -16,16 +17,31 @@ func improveResourcePluralization() PipelineStage {
 
 	return PipelineStage{
 		Name: "Improve resource pluralization",
-		Action: func(ctx context.Context, types Types) (Types, error) {
+		Action: func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
 
-			result := make(Types)
-			for typeName, typeDef := range types {
+			result := make(astmodel.Types)
+			var resources []astmodel.TypeDefinition
+			for _, typeDef := range types {
 				if _, ok := typeDef.Type().(*astmodel.ResourceType); ok {
-					newTypeName := typeName.Singular()
-					typeDef = typeDef.WithName(newTypeName)
-					result[newTypeName] = typeDef
+					resources = append(resources, typeDef)
 				} else {
-					result[typeName] = typeDef
+					result.Add(typeDef)
+				}
+			}
+
+			// now sort resources so that plurals will come after any singulars that
+			// already exist, so we don't conflict
+			sort.Slice(resources, func(i, j int) bool {
+				return resources[i].Name().Name() < resources[j].Name().Name()
+			})
+
+			for _, resource := range resources {
+				newTypeName := resource.Name().Singular()
+				if _, ok := result[newTypeName]; !ok {
```

Comment:
Rather than do this sorting can't you just check the provided `types`? If you `Singular()` something and it's in `types` then skip singularizing it?

I guess that would miss something weird like`Fooss` and `Foos` both being mapped to `Foo`? I'm not actually sure what happens when calling `Singular()` though so not sure that's even possible?

Reply:
hmm, I'm not sure why I didn't think of that